src/Microsoft.OData.Core/Uri/ODataUriSlim.cs

```diff
@@ -0,0 +1,139 @@
+﻿// <copyright file="ODataUriSlim.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Diagnostics;
+using System.Collections.Generic;
+using Microsoft.OData.UriParser.Aggregation;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// A lightweight version of <see cref="ODataUri"/> that is
+    /// used by the writer to efficiently create a copy of the URI when entering
+    /// each scope. This is used to avoid cloning <see cref="ODataUri"/>
+    /// excessively.
+    /// </summary>
+    internal struct ODataUriSlim
+    {
+        private readonly ODataUri odataUri;
+
+        /// <summary>
+        /// Creates an instance of <see cref="ODataUriSlim"/>
+        /// by copying the provided <see cref="ODataUri"/> instance.
+        /// </summary>
+        /// <param name="odataUri">The URI to copy.</param>
+        public ODataUriSlim(ODataUri odataUri)
+        {
+            Debug.Assert(odataUri != null, "odataUri != null");
+            this.odataUri = odataUri;
+            this.SelectAndExpand = odataUri.SelectAndExpand;
+            this.Path = odataUri.Path;
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="ODataUriSlim"/> by
+        /// cloning the provided <see cref="ODataUriSlim"/>.
+        /// </summary>
+        /// <param name="odataUriSlim">The URI to copy.</param>
+        public ODataUriSlim(ODataUriSlim odataUriSlim)
+        {
+            this.odataUri = odataUriSlim.odataUri;
+            this.SelectAndExpand = odataUriSlim.SelectAndExpand;
+            this.Path = odataUriSlim.Path;
+        }
+
+        /// <summary>
+        /// Gets or sets the top level path for this uri.
+        /// </summary>
+        public ODataPath Path { get; set; }
+
+        /// <summary>
+        /// Gets or sets any $select or $expand option for this uri.
+        /// </summary>
+        public SelectExpandClause SelectAndExpand { get; set; }
+
+        /// <summary>
+        /// Gets or sets the request Uri.

```

Comment:
```suggestion
        /// Gets the request Uri.
```
Justification: Property is read only